### PR TITLE
BC-3799 Display extended file info

### PR DESCRIPTION
**Pull request by dyedwiper**
# Short Description

## Links to Ticket and related Pull-Requests
https://ticketsystem.dbildungscloud.de/browse/BC-3799

## Checklist before merging

- [x] QA: In addition to review, the code has been manually tested (if manual testing is possible)
- [x] PO: Any deviation from requirements was agreed with Product-Owner / ticket author / support-team

> Notice: Please keep this Pull-Request as a Draft (or add WIP label), until it is ready to be reviewed

### CODE REVIEW

**Comment by bergatco** [comment]
Please add  tests for your new component `FileInfoChips` as well as extend the tests for `FileContentElement` to check for the parent component handing down the correct properties to its child component.

File: src/components/feature-board/shared/FileInfoChips.vue
Comment (bergatco) [inline src/components/feature-board/shared/FileInfoChips.vue:1]: Is it already decided to re-use this component somewhere else? Otherwise I would prefer to keep the component for now close to our `FileContentElement` and rename it to `FileContentElementChips` and move it out of the `shared` folder.

File: src/components/feature-board/shared/FileInfoChips.vue [hunk 1]
```diff
@@ -0,0 +1,55 @@
+<template>
+	<div>
```
Comment (SevenWaysDP) [inline src/components/feature-board/shared/FileInfoChips.vue:2]: Please use chip-groups

https://v2.vuetifyjs.com/en/components/chip-groups/#usage

File: src/components/feature-board/shared/FileInfoChips.vue [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+import { computed, defineComponent, PropType } from "vue";
+import { FileRecordResponse } from "@/fileStorageApi/v3";
+
+export default defineComponent({
+	name: "FileInfoChips",
+	props: {
+		fileRecord: {
+			type: Object as PropType<FileRecordResponse>,
+			required: true,
+		},
```
Comment (SevenWaysDP) [inline src/components/feature-board/shared/FileInfoChips.vue:18]: Please define the props you just need [fileSize, fileName]

File: src/components/feature-board/shared/FileInfoChips.vue [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+			while (convertedSize >= threshold && units.length >= power) {
+				convertedSize /= threshold;
+				unit = units[power];
+				power++;
+			}
+
+			const humanReadableFileSize = convertedSize.toFixed(2) + " " + unit;
+
+			return humanReadableFileSize;
+		});
```
Comment (SevenWaysDP) [inline src/components/feature-board/shared/FileInfoChips.vue:39]: Move the logic to a helper or Composition

**Review by SevenWaysDP — COMMENTED** [review]
Please create tests

File: src/components/feature-board/shared/FileInfoChips.vue (hunk 3 above)
Comment (dyedwiper) [reply src/components/feature-board/shared/FileInfoChips.vue:39]: We moved it in the utils folder.

File: src/components/feature-board/shared/FileInfoChips.vue
Comment (dyedwiper) [reply src/components/feature-board/shared/FileInfoChips.vue:1]: We thought about that too and weren't sure. To our knowledge it is not decided to reuse that. So we move it where you suggest.

File: src/components/feature-board/shared/FileInfoChips.vue (hunk 2 above)
Comment (SevenWaysDP) [reply src/components/feature-board/shared/FileInfoChips.vue]: It is now removed from FileContentElementEdit. For generic elements: if we have legacy file Element?

Comment (KalliSfak) [reply src/components/feature-board/shared/FileInfoChips.vue]: We defined the fileSize and fileName in props

File: src/components/feature-board/content-elements/FileContentElementChips.unit.ts
Comment (bergatco) [inline src/components/feature-board/content-elements/FileContentElementChips.unit.ts:1]: Please split up the tests! In this file please keep / add only tests related to the `FileContentElementChips` component and all of its displayed Parts. Move all the tests related to `getFileExtension` and `convertFileSizeToHumanReadable` to a new test file at `src/utils/fileHelper.unit.ts`
Further please also extend the tests in order to check, whether the chips do display the input data correctly

File: src/components/feature-board/content-elements/FileContentElementChips.unit.ts [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+			);
+		});
+
+		it("should return file size converted in GB", () => {
+			setup({ fileRecordSize: 1073741824, fileRecordName: "image.jpeg" });
+			expect(convertFileSizeToHumanReadable(1073741824)).toStrictEqual(
+				"1.00 GB"
+			);
+		});
+	});
```
Comment (SevenWaysDP) [inline src/components/feature-board/content-elements/FileContentElementChips.unit.ts:56]: This are not unit tests. Here two tests are enough, further tests are needed for fileHelper

File: src/components/feature-board/shared/FileInfoChips.vue (hunk 1 above)
Comment (dyedwiper) [reply src/components/feature-board/shared/FileInfoChips.vue]: Ich hab mir die Stelle nochmal angeguckt, weil ich mich gewundert hatte, dass man die Chips disablen muss, sie aber in der Doku nicht interaktiv sind: https://vuetifyjs.com/en/components/chips/ Turns out, die Interaktivität kommt von der Chip-Group: https://vuetifyjs.com/en/components/chip-groups/ Da steht auch: "It is used for creating groups of selections using chips." Und wir wollen hier ja keine Selection. Daher hab da wieder ein div hin gemacht.

File: src/utils/fileHelper.ts [hunk 5]
```diff
@@ -15,3 +15,35 @@ export async function dataUrlToFile(
...
+	let unit = units[0];
+	let power = 1;
+
+	while (convertedSize >= threshold && units.length >= power) {
+		convertedSize /= threshold;
+		unit = units[power];
+		power++;
+	}
+
+	const localeString = convertedSize.toLocaleString("de-DE", {
```
[the remark this answers is not in the document]
Comment (dyedwiper) [reply src/utils/fileHelper.ts]: Wir haben jetzt "window.navigator.language" benutzt, um die Sprache vom Browser auszulesen. Es gäbe ja auch noch die Möglichkeit die eingestellte Sprache aus der Bildungscloud zu nehmen. Wir haben uns gefragt, was besser ist, und waren unschlüssig...

[the remark this answers is not in the document]
Comment (SevenWaysDP) [reply src/utils/fileHelper.ts]: Kannst du diese Frage bei Frontändler stellen. Ich denke das wäre besser, wenn wir "eingestellte Sprache" uns richten. Aber ich habe keine Ahnung ob wir etwas übersehen.

**Review by SevenWaysDP — APPROVED** [review]

File: src/utils/fileHelper.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (bergatco) [reply src/utils/fileHelper.ts]: Die Nutzersprache wird in `nuxt-client/src/components/molecules/LanguageMenu.vue` gesetzt. Dazu wird die Funktion `updateUserLanguage` des `AuthModule` (`nuxt-client/src/store/auth.ts`) aufgerufen. Demnach wird die `locale` im `AuthModule` dann gesetzt. Dem entsprechend könnte man eventuell die Funktion `getLocale` verwenden, um die aktuelle Sprache zu bestimmen?

**Review by KalliSfak — DISMISSED** [review]

File: src/components/feature-board/content-elements/FileContentElementChips.vue [hunk 6]
```diff
@@ -0,0 +1,43 @@
+<template>
+	<div class="ma-2">
```
Comment (bergatco) [inline src/components/feature-board/content-elements/FileContentElementChips.vue:2]: I just did check the styling again the designs in Figma. Please use `padding` instead of a `margin` here.

```suggestion
	<div class="pa-2">
```

File: src/components/feature-board/content-elements/FileContentElementChips.vue [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+		<v-chip class="mr-2" small>{{ fileExtension }}</v-chip>
+		<v-chip class="mr-2" small>{{ humanReadableFileSize }}</v-chip>
```
Comment (bergatco) [inline src/components/feature-board/content-elements/FileContentElementChips.vue:4]: I also found, that the `v-chips` are still having an `hover` effect. Please apply the changes suggested by @SevenWaysDP .

```suggestion
		<v-chip class="mr-2" disabled small>{{ fileExtension }}</v-chip>
		<v-chip class="mr-2" disabled small>{{ humanReadableFileSize }}</v-chip>
```

File: src/components/feature-board/content-elements/FileContentElementChips.vue [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+			getFileExtension(props.fileName).toUpperCase()
+		);
+
+		return {
+			humanReadableFileSize,
+			fileExtension,
+		};
+	},
+});
+</script>
```
Comment (bergatco) [inline src/components/feature-board/content-elements/FileContentElementChips.vue:43]: I also found, that the `v-chips` are still having an `hover` effect. Please apply the changes suggested by @SevenWaysDP .

```suggestion
</script>

<style lang="scss" scoped>
.v-chip {
	opacity: 1;
}
</style>

```

File: src/components/feature-board/content-elements/FileContentElementChips.vue
Comment (bergatco) [inline src/components/feature-board/content-elements/FileContentElementChips.vue:1]: Otherwise the styling does look good and is according to the mocks. 👍

**Review by bergatco — CHANGES_REQUESTED** [review]

File: src/components/feature-board/content-elements/FileContentElementChips.vue (hunk 6 above)
Comment (dyedwiper) [reply src/components/feature-board/content-elements/FileContentElementChips.vue:2]: I didn't know that you can look it up in Figma. Interesting! It won't make any visible difference at the moment. This would only be the case if we had a different background, but this might change sometime... So we will use padding.

File: src/components/feature-board/content-elements/FileContentElementChips.vue (hunk 7 above)
Comment (dyedwiper) [reply src/components/feature-board/content-elements/FileContentElementChips.vue:4]: I'm wondering where this hover effect comes from because a v-chip doesn't normally have it. What I found is that when hovering the following pseudo-element is rendered: ".theme--light.v-chip:hover::before" with style "opacity: 0.04;". Do you have any idea where that comes from?

File: src/components/feature-board/content-elements/FileContentElement.unit.ts
Comment (bergatco) [inline src/components/feature-board/content-elements/FileContentElement.unit.ts:1]: Please also add tests for other scenarios :

-  "when the file needs to be uploaded" & "when upload is successful", the chips should be visible in both modes, edit & view.
-  "when the file needs to be uploaded" & "when upload throws error", the chips should NOT be visible in both modes, edit & view.
- "when file is uploaded" & "when a virus is detected", the chips should be visible in both modes, edit & view.

File: src/components/feature-board/content-elements/FileContentElementChips.vue (hunk 7 above)
Comment (bergatco) [reply src/components/feature-board/content-elements/FileContentElementChips.vue:4]: Unfortunately, I don't. I also found while doing some previous research (see [Colored section of Chips in vuetify docs](https://v2.vuetifyjs.com/en/components/chips/#colored) , that the "default" chip does have a hover effect. If you change the `color` of it, the hover effect disappears.
But as we are using the "default" color here, please implement the workaround kindly provided by @SevenWaysDP.

File: src/components/feature-board/content-elements/FileContentElementChips.unit.ts
Comment (bergatco) [inline src/components/feature-board/content-elements/FileContentElementChips.unit.ts:1]: Maybe it would also be a good idea to add some test cases for edge cases, like when either file size or name are not available (undefined)?

File: src/utils/fileHelper.ts [hunk 9]
```diff
@@ -1,3 +1,5 @@
+import { authModule } from "@/store";
```
Comment (bergatco) [inline src/utils/fileHelper.ts:1]: This line can be removed, as `authModule` is not used anywhere in the file.

File: src/utils/fileHelper.unit.ts
Comment (bergatco) [inline src/utils/fileHelper.unit.ts:1]: Please add also some edge cases, when no or wrong data is provided to both functions.

File: src/utils/fileHelper.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (KalliSfak) [reply src/utils/fileHelper.ts]: We used the i18n.n function for the localization.

**Review by bergatco — DISMISSED** [review]

**Review by bergatco — APPROVED** [review]
